compiler/codegen/OMRPeephole.cpp

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "codegen/Peephole.hpp"
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/Instruction.hpp"
+
+OMR::Peephole::Peephole(TR::Compilation* comp) :
+   _comp(comp), _cg(comp->cg())
+   {}
+
+TR::Peephole*
+OMR::Peephole::self()
+   {
+   return static_cast<TR::Peephole*>(this);
+   }
+
+TR::Compilation*
+OMR::Peephole::comp() const
+   {
+   return _comp;
+   }
+
+TR::CodeGenerator*
+OMR::Peephole::cg() const
+   {
+   return _cg;
+   }
+
+bool
+OMR::Peephole::perform()
+   {
+   bool performed = false;
+
+   TR::Instruction* prevInst = NULL;
+   TR::Instruction* currInst = self()->cg()->getFirstInstruction();
+
+   while (currInst != NULL)
+      {
+      performed |= self()->performOnInstruction(currInst);
+
+      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
+      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
+      if (currInst->getPrev() == prevInst)
+         {
+         prevInst = currInst;
+         currInst = currInst->getNext();
+         }
+      else
+         {
+         currInst = prevInst->getNext();
+         }
+      }
```

Comment:
How much is "earlier"? Right now if any transformation is performed, i.e. even if some further down instruction is modified, we go back one instruction and restart the peephole there. So for example say we had the following sequence of instructions:

```
X1 
X2 <-- Currently about to call `performOnInstruction` on this one
X3
X4
```

Say `performOnInstruction` returned true, and the instruction sequence following the return looks like:

```
X1
X2
X3
Z4 <-- Got modified to Z4
```

Because `performOnInstruction` returned true, the above code would restart the peephole window at `X1` rather than `X3`, irregardless of what transformation was actually performed. This is because peephole windows only slide down, and peephole should only be modifying instructions following the current instruction being examined, never before!

This should cover the case where new opportunities are unlocked, and it also shields us from `X2` being completely removed from the instruction stream for example.

Reply:
I don't think going back one instruction is sufficient in all cases, though I concede that it may work given our current set of peepholes. Imagine a theoretical situation with the following peepholes:

1. Any time the sequence `C`, `D` shows up, replace it with `E`
2. Any time the sequence `A`, `B`, `E` shows up, replace it with `F`

Now imagine running those peepholes on the following set of instructions:

```
--> A
    B
    C
    D
```

When we first reach `A`, peephole 2 cannot be used, since the instruction after the next one is `C`, not `E`. We then reach `C` and apply peephole 1 to turn the sequence into:

```
    A
--> B
    E
```

However, by only going back one instruction, we don't ever actually end up going back to `A` and thus miss the opportunity to replace the entire sequence with a single instruction `F`.

Reply:
Right for sure there will be missed opportunities. Same issue happens with the simplifier. The way we overcome that in the optimizer is by running the opt multiple times. In general there will be no way to know how far back we have to go to ensure we catch the opportunity you describe because that would involve peephole 1. being aware of peephole 2., and in general all other peepholes as well. This doesn't seem feasible.

We can run the peephole iteratively until the `perform()` function returns false, however the complexity of such algorithm is at least O(n^2) where n is the number of instructions. IMO the peephole pass should be quick and simple reductions. We would probably catch a large majority of cases by running peepholes twice if need be.

Reply:
Fair enough, I'm fine with whatever we come up with here (even if it's not perfect) as long as the issue has been considered and what we have handles the majority of the cases we want to optimize. You'd need to either keep running peepholes until nothing happens or back up by the maximum window size of a peephole to fully solve the problem, which may not be an acceptable cost. This was just a thought I had while reading the code after it was moved.